File: python/command/commands/remote_control.py
from commands.command import Command, NoCommand


class RemoteControl:

    def __init__(self):
        self.on_commands = []
        self.off_commands = []
        
        no_command = NoCommand()

        for _ in range(2):
            self.on_commands.append(no_command)
            self.off_commands.append(no_command)

        self.undo_command = no_command

    def set_command(self, slot: int, on_command: Command, off_command: Command) -> None:
        self.on_commands[slot] = on_command
        self.off_commands[slot] = off_command

    def on_button_was_pushed(self, slot: int) -> None:
        self.on_commands[slot].execute()
        self.undo_command = self.on_commands[slot]

    def off_button_was_pushed(self, slot: int) -> None:
        self.off_commands[slot].execute()
        self.undo_command = self.off_commands[slot]

    def undo_was_pushed(self) -> None:
        self.undo_command.undo()

File: java/patterns_example/src/main/java/com/app/patterns/command/RemoteControl.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.app.patterns.command;

import com.app.patterns.command.Command;


public class RemoteControl {
    
    Command[] onCommands;
    Command[] offCommands;
    Command undoCommand;
    
    public RemoteControl() {
        
        onCommands = new Command[3];
        offCommands = new Command[3];
        Command noCommand = new NoCommand();
        
        for (int i = 0; i < offCommands.length ; i++) {
            onCommands[i] = noCommand;
            offCommands[i] = noCommand;
        }
        
        undoCommand = noCommand;
    }
    
    public void setCommand(int slot, Command onCommand, Command offCommand) {
        onCommands[slot] = onCommand;
        offCommands[slot] = offCommand;
    }
    
    public void onButtonWasPushed(int slot) {
        onCommands[slot].execute();
        undoCommand = onCommands[slot];
    }
    
    public void offButtonWasPushed(int slot) {
        offCommands[slot].execute();
        undoCommand = offCommands[slot];
    }
    
    public void undoButtonWasPushed() {
        undoCommand.undo();
    }    
    
    @Override
    public String toString() {
        
        StringBuilder stringBuff = new StringBuilder();
        stringBuff.append("\n------ Remote Control -------\n");
        
        for (int i = 0; i < onCommands.length; i++) {
            stringBuff.append("[slot ")
                    .append(i)
                    .append("] ")
                    .append(onCommands[i].getClass().getName())
                    .append(" ")
                    .append(offCommands[i].getClass().getName())
                    .append("\n");
        }
        
        return stringBuff.toString();
    }
    
}
